File: src/ast/TypeScript.zig
// This function is taken from the official TypeScript compiler source code:
// https://github.com/microsoft/TypeScript/blob/master/src/compiler/parser.ts
pub fn canFollowTypeArgumentsInExpression(p: anytype) bool {
    return switch (p.lexer.token) {
        // These are the only tokens can legally follow a type argument list. So we
        // definitely want to treat them as type arg lists.
        .t_open_paren, // foo<x>(
        .t_no_substitution_template_literal, // foo<T> `...`
        // foo<T> `...${100}...`
        .t_template_head,
        => true,

        // A type argument list followed by `<` never makes sense, and a type argument list followed
        // by `>` is ambiguous with a (re-scanned) `>>` operator, so we disqualify both. Also, in
        // this context, `+` and `-` are unary operators, not binary operators.
        .t_less_than,
        .t_greater_than,
        .t_plus,
        .t_minus,
        // TypeScript always sees "t_greater_than" instead of these tokens since
        // their scanner works a little differently than our lexer. So since
        // "t_greater_than" is forbidden above, we also forbid these too.
        .t_greater_than_equals,
        .t_greater_than_greater_than,
        .t_greater_than_greater_than_equals,
        .t_greater_than_greater_than_greater_than,
        .t_greater_than_greater_than_greater_than_equals,
        => false,

        // We favor the type argument list interpretation when it is immediately followed by
        // a line break, a binary operator, or something that can't start an expression.
        else => p.lexer.has_newline_before or isBinaryOperator(p) or !isStartOfExpression(p),
    };
}

pub const Metadata = union(enum) {
    m_none: void,

    m_never: void,
    m_unknown: void,
    m_any: void,
    m_void: void,
    m_null: void,
    m_undefined: void,
    m_function: void,
    m_array: void,
    m_boolean: void,
    m_string: void,
    m_object: void,
    m_number: void,
    m_bigint: void,
    m_symbol: void,
    m_promise: void,
    m_identifier: Ref,
    m_dot: List(Ref),

    pub const default: @This() = .m_none;

    // the logic in finishUnion, mergeUnion, finishIntersection and mergeIntersection is
    // translated from:
    // https://github.com/microsoft/TypeScript/blob/e0a324b0503be479f2b33fd2e17c6e86c94d1297/src/compiler/transformers/typeSerializer.ts#L402

    /// Return the final union type if possible, or return null to continue merging.
    ///
    /// If the current type is m_never, m_null, or m_undefined assign the current type
    /// to m_none and return null to ensure it's always replaced by the next type.
    pub fn finishUnion(current: *@This(), p: anytype) ?@This() {
        return switch (current.*) {
            .m_identifier => |ref| {
                if (strings.eqlComptime(p.loadNameFromRef(ref), "Object")) {
                    return .m_object;
                }
                return null;
            },

            .m_unknown,
            .m_any,
            .m_object,
            => .m_object,

            .m_never,
            .m_null,
            .m_undefined,
            => {
                current.* = .m_none;
                return null;
            },

            else => null,
        };
    }

    pub fn mergeUnion(result: *@This(), left: @This()) void {
        if (left != .m_none) {
            if (std.meta.activeTag(result.*) != std.meta.activeTag(left)) {
                result.* = switch (result.*) {
                    .m_never,
                    .m_undefined,
                    .m_null,
                    => left,

                    else => .m_object,
                };
            } else {
                switch (result.*) {
                    .m_identifier => |ref| {
                        if (!ref.eql(left.m_identifier)) {
                            result.* = .m_object;
                        }
                    },
                    else => {},
                }
            }
        } else {
            // always take the next value if left is m_none
        }
    }

    /// Return the final intersection type if possible, or return null to continue merging.
    ///
    /// If the current type is m_unknown, m_null, or m_undefined assign the current type
    /// to m_none and return null to ensure it's always replaced by the next type.
    pub fn finishIntersection(current: *@This(), p: anytype) ?@This() {
        return switch (current.*) {
            .m_identifier => |ref| {
                if (strings.eqlComptime(p.loadNameFromRef(ref), "Object")) {
                    return .m_object;
                }
                return null;
            },

            // ensure m_never is the final type
            .m_never => .m_never,

            .m_any,
            .m_object,
            => .m_object,

            .m_unknown,
            .m_null,
            .m_undefined,
            => {
                current.* = .m_none;
                return null;
            },

            else => null,
        };
    }

    pub fn mergeIntersection(result: *@This(), left: @This()) void {
        if (left != .m_none) {
            if (std.meta.activeTag(result.*) != std.meta.activeTag(left)) {
                result.* = switch (result.*) {
                    .m_unknown,
                    .m_undefined,
                    .m_null,
                    => left,

                    // ensure m_never is the final type
                    .m_never => .m_never,

                    else => .m_object,
                };
            } else {
                switch (result.*) {
                    .m_identifier => |ref| {
                        if (!ref.eql(left.m_identifier)) {
                            result.* = .m_object;
                        }
                    },
                    else => {},
                }
            }
        } else {
            // make sure intersection of only m_unknown serializes to "undefined"
            // instead of "Object"
            if (result.* == .m_unknown) {
                result.* = .m_undefined;
            }
        }
    }
};

pub fn isTSArrowFnJSX(p: anytype) !bool {
    const old_lexer = p.lexer;

    try p.lexer.next();
    // Look ahead to see if this should be an arrow function instead
    var is_ts_arrow_fn = false;

    if (p.lexer.token == .t_const) {
        try p.lexer.next();
    }
    if (p.lexer.token == .t_identifier) {
        try p.lexer.next();
        if (p.lexer.token == .t_comma or p.lexer.token == .t_equals) {
            is_ts_arrow_fn = true;
        } else if (p.lexer.token == .t_extends) {
            try p.lexer.next();
            is_ts_arrow_fn = p.lexer.token != .t_equals and p.lexer.token != .t_greater_than and p.lexer.token != .t_slash;
        }
    }

    // Restore the lexer
    p.lexer.restore(&old_lexer);
    return is_ts_arrow_fn;
}

// This function is taken from the official TypeScript compiler source code:
// https://github.com/microsoft/TypeScript/blob/master/src/compiler/parser.ts
fn isBinaryOperator(p: anytype) bool {
    return switch (p.lexer.token) {
        .t_in => p.allow_in,

        .t_question_question,
        .t_bar_bar,
        .t_ampersand_ampersand,
        .t_bar,
        .t_caret,
        .t_ampersand,
        .t_equals_equals,
        .t_exclamation_equals,
        .t_equals_equals_equals,
        .t_exclamation_equals_equals,
        .t_less_than,
        .t_greater_than,
        .t_less_than_equals,
        .t_greater_than_equals,
        .t_instanceof,
        .t_less_than_less_than,
        .t_greater_than_greater_than,
        .t_greater_than_greater_than_greater_than,
        .t_plus,
        .t_minus,
        .t_asterisk,
        .t_slash,
        .t_percent,
        .t_asterisk_asterisk,
        => true,
        .t_identifier => p.lexer.isContextualKeyword("as") or p.lexer.isContextualKeyword("satisfies"),
        else => false,
    };
}

// This function is taken from the official TypeScript compiler source code:
// https://github.com/microsoft/TypeScript/blob/master/src/compiler/parser.ts
fn isStartOfLeftHandSideExpression(p: anytype) bool {
    return switch (p.lexer.token) {
        .t_this,
        .t_super,
        .t_null,
        .t_true,
        .t_false,
        .t_numeric_literal,
        .t_big_integer_literal,
        .t_string_literal,
        .t_no_substitution_template_literal,
        .t_template_head,
        .t_open_paren,
        .t_open_bracket,
        .t_open_brace,
        .t_function,
        .t_class,
        .t_new,
        .t_slash,
        .t_slash_equals,
        .t_identifier,
        => true,
        .t_import => lookAheadNextTokenIsOpenParenOrLessThanOrDot(p),
        else => isIdentifier(p),
    };
}

fn lookAheadNextTokenIsOpenParenOrLessThanOrDot(p: anytype) bool {
    const old_lexer = p.lexer;
    const old_log_disabled = p.lexer.is_log_disabled;
    p.lexer.is_log_disabled = true;
    defer {
        p.lexer.restore(&old_lexer);
        p.lexer.is_log_disabled = old_log_disabled;
    }
    p.lexer.next() catch {};

    return switch (p.lexer.token) {
        .t_open_paren, .t_less_than, .t_dot => true,
        else => false,
    };
}

// This function is taken from the official TypeScript compiler source code:
// https://github.com/microsoft/TypeScript/blob/master/src/compiler/parser.ts
fn isIdentifier(p: anytype) bool {
    if (p.lexer.token == .t_identifier) {
        // If we have a 'yield' keyword, and we're in the [yield] context, then 'yield' is
        // considered a keyword and is not an identifier.
        if (p.fn_or_arrow_data_parse.allow_yield != .allow_ident and strings.eqlComptime(p.lexer.identifier, "yield")) {
            return false;
        }

        // If we have an 'await' keyword, and we're in the [await] context, then 'await' is
        // considered a keyword and is not an identifier.
        if (p.fn_or_arrow_data_parse.allow_await != .allow_ident and strings.eqlComptime(p.lexer.identifier, "await")) {
            return false;
        }

        return true;
    }

    return false;
}

fn isStartOfExpression(p: anytype) bool {
    if (isStartOfLeftHandSideExpression(p))
        return true;

    switch (p.lexer.token) {
        .t_plus,
        .t_minus,
        .t_tilde,
        .t_exclamation,
        .t_delete,
        .t_typeof,
        .t_void,
        .t_plus_plus,
        .t_minus_minus,
        .t_less_than,
        .t_private_identifier,
        .t_at,
        => return true,
        else => {
            if (p.lexer.token == .t_identifier and (strings.eqlComptime(p.lexer.identifier, "await") or strings.eqlComptime(p.lexer.identifier, "yield"))) {
                // Yield/await always starts an expression.  Either it is an identifier (in which case
                // it is definitely an expression).  Or it's a keyword (either because we're in
                // a generator or async function, or in strict mode (or both)) and it started a yield or await expression.
                return true;
            }

            // Error tolerance.  If we see the start of some binary operator, we consider
            // that the start of an expression.  That way we'll parse out a missing identifier,
            // give a good message about an identifier being missing, and then consume the
            // rest of the binary expression.
            if (isBinaryOperator(p)) {
                return true;
            }

            return isIdentifier(p);
        },
    }

    unreachable;
}

pub const Identifier = struct {
    pub const StmtIdentifier = enum {
        s_type,

        s_namespace,

        s_abstract,

        s_module,

        s_interface,

        s_declare,
    };
    pub fn forStr(str: string) ?StmtIdentifier {
        switch (str.len) {
            "type".len => return if (strings.eqlComptimeIgnoreLen(str, "type"))
                .s_type
            else
                null,
            "interface".len => {
                if (strings.eqlComptime(str, "interface")) {
                    return .s_interface;
                } else if (strings.eqlComptime(str, "namespace")) {
                    return .s_namespace;
                } else {
                    return null;
                }
            },
            "abstract".len => {
                if (strings.eqlComptime(str, "abstract")) {
                    return .s_abstract;
                } else {
                    return null;
                }
            },
            "declare".len => {
                if (strings.eqlComptime(str, "declare")) {
                    return .s_declare;
                } else {
                    return null;
                }
            },
            "module".len => {
                if (strings.eqlComptime(str, "module")) {
                    return .s_module;
                } else {
                    return null;
                }
            },
            else => return null,
        }
    }
    pub const IMap = bun.ComptimeStringMap(Kind, .{
        .{ "unique", .unique },
        .{ "abstract", .abstract },
        .{ "asserts", .asserts },

        .{ "keyof", .prefix_keyof },
        .{ "readonly", .prefix_readonly },

        .{ "any", .primitive_any },
        .{ "never", .primitive_never },
        .{ "unknown", .primitive_unknown },
        .{ "undefined", .primitive_undefined },
        .{ "object", .primitive_object },
        .{ "number", .primitive_number },
        .{ "string", .primitive_string },
        .{ "boolean", .primitive_boolean },
        .{ "bigint", .primitive_bigint },
        .{ "symbol", .primitive_symbol },

        .{ "infer", .infer },
    });
    pub const Kind = enum {
        normal,
        unique,
        abstract,
        asserts,
        prefix_keyof,
        prefix_readonly,
        primitive_any,
        primitive_never,
        primitive_unknown,
        primitive_undefined,
        primitive_object,
        primitive_number,
        primitive_string,
        primitive_boolean,
        primitive_bigint,
        primitive_symbol,
        infer,
    };
};

pub const SkipTypeOptions = enum {
    is_return_type,
    is_index_signature,
    allow_tuple_labels,
    disallow_conditional_types,

    pub const Bitset = std.enums.EnumSet(@This());
    pub const empty = Bitset.initEmpty();
};

const string = []const u8;

const bun = @import("bun");
const strings = bun.strings;

const js_lexer = bun.js_lexer;
const T = js_lexer.T;

const js_parser = bun.js_parser;
const Ref = js_parser.Ref;
const TypeScript = js_parser.TypeScript;

const std = @import("std");
const List = std.ArrayListUnmanaged;

File: src/js_parser/ast/TypeScript.rs
use crate::ast::p::P;
use crate::js_lexer::T;
use crate::parser::JsxT;
use crate::Ref;

// Zig: `p: anytype` for the generic parser instance. Round-C lowered NewParser_ →
// `P<'a, const TS, J: JsxT, const SCAN>`. The Phase-A draft used unbounded `<P>` which
// can't access fields; convert to `impl P` methods. The `Metadata::*` methods that need
// `p.load_name_from_ref` take a closure to avoid the impl-on-foreign-type problem.

// This function is taken from the official TypeScript compiler source code:
// https://github.com/microsoft/TypeScript/blob/master/src/compiler/parser.ts
impl<'a, const TS: bool, J: JsxT, const SCAN: bool> P<'a, TS, J, SCAN> {
pub fn can_follow_type_arguments_in_expression(&mut self) -> bool {
    let p = self;
    match p.lexer.token {
        // These are the only tokens can legally follow a type argument list. So we
        // definitely want to treat them as type arg lists.
        T::TOpenParen // foo<x>(
        | T::TNoSubstitutionTemplateLiteral // foo<T> `...`
        // foo<T> `...${100}...`
        | T::TTemplateHead => true,

        // A type argument list followed by `<` never makes sense, and a type argument list followed
        // by `>` is ambiguous with a (re-scanned) `>>` operator, so we disqualify both. Also, in
        // this context, `+` and `-` are unary operators, not binary operators.
        T::TLessThan
        | T::TGreaterThan
        | T::TPlus
        | T::TMinus
        // TypeScript always sees "t_greater_than" instead of these tokens since
        // their scanner works a little differently than our lexer. So since
        // "t_greater_than" is forbidden above, we also forbid these too.
        | T::TGreaterThanEquals
        | T::TGreaterThanGreaterThan
        | T::TGreaterThanGreaterThanEquals
        | T::TGreaterThanGreaterThanGreaterThan
        | T::TGreaterThanGreaterThanGreaterThanEquals => false,

        // We favor the type argument list interpretation when it is immediately followed by
        // a line break, a binary operator, or something that can't start an expression.
        _ => p.lexer.has_newline_before || p.is_binary_operator() || !p.is_start_of_expression(),
    }
}
} // end impl P (can_follow_type_arguments_in_expression)

#[derive(Clone)]
pub enum Metadata {
    MNone,

    MNever,
    MUnknown,
    MAny,
    MVoid,
    MNull,
    MUndefined,
    MFunction,
    MArray,
    MBoolean,
    MString,
    MObject,
    MNumber,
    MBigint,
    MSymbol,
    MPromise,
    MIdentifier(Ref),
    // TODO(port): Zig used `std.ArrayListUnmanaged(Ref)`. This is an AST crate;
    // if this list is arena-backed in practice, switch to
    // `bun_alloc::ArenaVec<'bump, Ref>` in Phase B.
    MDot(Vec<Ref>),
}

impl Default for Metadata {
    fn default() -> Self {
        Metadata::MNone
    }
}

impl Metadata {
    pub const DEFAULT: Self = Metadata::MNone;

    // the logic in finish_union, merge_union, finish_intersection and merge_intersection is
    // translated from:
    // https://github.com/microsoft/TypeScript/blob/e0a324b0503be479f2b33fd2e17c6e86c94d1297/src/compiler/transformers/typeSerializer.ts#L402

    /// Return the final union type if possible, or return None to continue merging.
    ///
    /// If the current type is MNever, MNull, or MUndefined assign the current type
    /// to MNone and return None to ensure it's always replaced by the next type.
    /// `load_name`: closure form of `p.load_name_from_ref` to avoid coupling Metadata to P.
    pub fn finish_union<'b, F: Fn(Ref) -> &'b [u8]>(&mut self, load_name: F) -> Option<Self> {
        let current = self;
        match current {
            Metadata::MIdentifier(r) => {
                if load_name(*r) == b"Object" {
                    return Some(Metadata::MObject);
                }
                None
            }

            Metadata::MUnknown | Metadata::MAny | Metadata::MObject => Some(Metadata::MObject),

            Metadata::MNever | Metadata::MNull | Metadata::MUndefined => {
                *current = Metadata::MNone;
                None
            }

            _ => None,
        }
    }

    pub fn merge_union(&mut self, left: Self) {
        let result = self;
        if !matches!(left, Metadata::MNone) {
            if core::mem::discriminant(result) != core::mem::discriminant(&left) {
                *result = match result {
                    Metadata::MNever | Metadata::MUndefined | Metadata::MNull => left,

                    _ => Metadata::MObject,
                };
            } else {
                // PORT NOTE: reshaped for borrowck — copy Ref out before reassigning *result
                if let Metadata::MIdentifier(r) = result {
                    let r = *r;
                    if let Metadata::MIdentifier(l) = left {
                        if !r.eql(l) {
                            *result = Metadata::MObject;
                        }
                    }
                }
            }
        } else {
            // always take the next value if left is MNone
        }
    }

    /// Return the final intersection type if possible, or return None to continue merging.
    ///
    /// If the current type is MUnknown, MNull, or MUndefined assign the current type
    /// to MNone and return None to ensure it's always replaced by the next type.
    pub fn finish_intersection<'b, F: Fn(Ref) -> &'b [u8]>(&mut self, load_name: F) -> Option<Self> {
        let current = self;
        match current {
            Metadata::MIdentifier(r) => {
                if load_name(*r) == b"Object" {
                    return Some(Metadata::MObject);
                }
                None
            }

            // ensure MNever is the final type
            Metadata::MNever => Some(Metadata::MNever),

            Metadata::MAny | Metadata::MObject => Some(Metadata::MObject),

            Metadata::MUnknown | Metadata::MNull | Metadata::MUndefined => {
                *current = Metadata::MNone;
                None
            }

            _ => None,
        }
    }

    pub fn merge_intersection(&mut self, left: Self) {
        let result = self;
        if !matches!(left, Metadata::MNone) {
            if core::mem::discriminant(result) != core::mem::discriminant(&left) {
                *result = match result {
                    Metadata::MUnknown | Metadata::MUndefined | Metadata::MNull => left,

                    // ensure MNever is the final type
                    Metadata::MNever => Metadata::MNever,

                    _ => Metadata::MObject,
                };
            } else {
                // PORT NOTE: reshaped for borrowck — copy Ref out before reassigning *result
                if let Metadata::MIdentifier(r) = result {
                    let r = *r;
                    if let Metadata::MIdentifier(l) = left {
                        if !r.eql(l) {
                            *result = Metadata::MObject;
                        }
                    }
                }
            }
        } else {
            // make sure intersection of only MUnknown serializes to "undefined"
            // instead of "Object"
            if matches!(result, Metadata::MUnknown) {
                *result = Metadata::MUndefined;
            }
        }
    }
}

impl<'a, const TS: bool, J: JsxT, const SCAN: bool> P<'a, TS, J, SCAN> {
// TODO(port): narrow error set — only `lexer.next()` is fallible here.
pub fn is_ts_arrow_fn_jsx(&mut self) -> Result<bool, bun_core::Error> {
    let p = self;
    // PORT NOTE: Zig `const old = p.lexer` (value copy). Rust Lexer holds `&mut Log`
    // so cannot Clone; use the LexerSnapshot POD via `snapshot()`/`restore()`.
    let old_lexer = p.lexer.snapshot();

    p.lexer.next()?;
    // Look ahead to see if this should be an arrow function instead
    let mut is_ts_arrow_fn = false;

    if p.lexer.token == T::TConst {
        p.lexer.next()?;
    }
    if p.lexer.token == T::TIdentifier {
        p.lexer.next()?;
        if p.lexer.token == T::TComma || p.lexer.token == T::TEquals {
            is_ts_arrow_fn = true;
        } else if p.lexer.token == T::TExtends {
            p.lexer.next()?;
            is_ts_arrow_fn = p.lexer.token != T::TEquals
                && p.lexer.token != T::TGreaterThan
                && p.lexer.token != T::TSlash;
        }
    }

    // Restore the lexer
    p.lexer.restore(&old_lexer);
    Ok(is_ts_arrow_fn)
}

// This function is taken from the official TypeScript compiler source code:
// https://github.com/microsoft/TypeScript/blob/master/src/compiler/parser.ts
fn is_binary_operator(&self) -> bool {
    let p = self;
    match p.lexer.token {
        T::TIn => p.allow_in,

        T::TQuestionQuestion
        | T::TBarBar
        | T::TAmpersandAmpersand
        | T::TBar
        | T::TCaret
        | T::TAmpersand
        | T::TEqualsEquals
        | T::TExclamationEquals
        | T::TEqualsEqualsEquals
        | T::TExclamationEqualsEquals
        | T::TLessThan
        | T::TGreaterThan
        | T::TLessThanEquals
        | T::TGreaterThanEquals
        | T::TInstanceof
        | T::TLessThanLessThan
        | T::TGreaterThanGreaterThan
        | T::TGreaterThanGreaterThanGreaterThan
        | T::TPlus
        | T::TMinus
        | T::TAsterisk
        | T::TSlash
        | T::TPercent
        | T::TAsteriskAsterisk => true,
        T::TIdentifier => {
            p.lexer.is_contextual_keyword(b"as") || p.lexer.is_contextual_keyword(b"satisfies")
        }
        _ => false,
    }
}

// This function is taken from the official TypeScript compiler source code:
// https://github.com/microsoft/TypeScript/blob/master/src/compiler/parser.ts
fn is_start_of_left_hand_side_expression(&mut self) -> bool {
    let p = self;
    match p.lexer.token {
        T::TThis
        | T::TSuper
        | T::TNull
        | T::TTrue
        | T::TFalse
        | T::TNumericLiteral
        | T::TBigIntegerLiteral
        | T::TStringLiteral
        | T::TNoSubstitutionTemplateLiteral
        | T::TTemplateHead
        | T::TOpenParen
        | T::TOpenBracket
        | T::TOpenBrace
        | T::TFunction
        | T::TClass
        | T::TNew
        | T::TSlash
        | T::TSlashEquals
        | T::TIdentifier => true,
        T::TImport => p.look_ahead_next_token_is_open_paren_or_less_than_or_dot(),
        _ => p.ts_is_identifier(),
    }
}

fn look_ahead_next_token_is_open_paren_or_less_than_or_dot(&mut self) -> bool {
    let p = self;
    // PORT NOTE: Zig value-copied the Lexer; use snapshot()/restore() (see is_ts_arrow_fn_jsx).
    let old_lexer = p.lexer.snapshot();
    let old_log_disabled = p.lexer.is_log_disabled;
    p.lexer.is_log_disabled = true;

    let _ = p.lexer.next();

    let result = matches!(p.lexer.token, T::TOpenParen | T::TLessThan | T::TDot);

    // PORT NOTE: Zig used `defer` for restore; reshaped to linear since there is
    // no early return between the defer and end of scope.
    p.lexer.restore(&old_lexer);
    p.lexer.is_log_disabled = old_log_disabled;

    result
}

// This function is taken from the official TypeScript compiler source code:
// https://github.com/microsoft/TypeScript/blob/master/src/compiler/parser.ts
// PORT NOTE: renamed `ts_is_identifier` to avoid clash with lexer/P helpers of the same name.
fn ts_is_identifier(&self) -> bool {
    use crate::parser::AwaitOrYield::AllowIdent;
    let p = self;
    if p.lexer.token == T::TIdentifier {
        // If we have a 'yield' keyword, and we're in the [yield] context, then 'yield' is
        // considered a keyword and is not an identifier.
        if p.fn_or_arrow_data_parse.allow_yield != AllowIdent && p.lexer.identifier == b"yield" {
            return false;
        }

        // If we have an 'await' keyword, and we're in the [await] context, then 'await' is
        // considered a keyword and is not an identifier.
        if p.fn_or_arrow_data_parse.allow_await != AllowIdent && p.lexer.identifier == b"await" {
            return false;
        }

        return true;
    }

    false
}

fn is_start_of_expression(&mut self) -> bool {
    let p = self;
    if p.is_start_of_left_hand_side_expression() {
        return true;
    }

    match p.lexer.token {
        T::TPlus
        | T::TMinus
        | T::TTilde
        | T::TExclamation
        | T::TDelete
        | T::TTypeof
        | T::TVoid
        | T::TPlusPlus
        | T::TMinusMinus
        | T::TLessThan
        | T::TPrivateIdentifier
        | T::TAt => true,
        _ => {
            if p.lexer.token == T::TIdentifier
                && (p.lexer.identifier == b"await" || p.lexer.identifier == b"yield")
            {
                // Yield/await always starts an expression.  Either it is an identifier (in which case
                // it is definitely an expression).  Or it's a keyword (either because we're in
                // a generator or async function, or in strict mode (or both)) and it started a yield or await expression.
                return true;
            }

            // Error tolerance.  If we see the start of some binary operator, we consider
            // that the start of an expression.  That way we'll parse out a missing identifier,
            // give a good message about an identifier being missing, and then consume the
            // rest of the binary expression.
            if p.is_binary_operator() {
                return true;
            }

            p.ts_is_identifier()
        }
    }
}
} // end impl P (predicate fns)

pub mod identifier {
    pub enum StmtIdentifier {
        SType,

        SNamespace,

        SAbstract,

        SModule,

        SInterface,

        SDeclare,
    }

    pub fn for_str(str: &[u8]) -> Option<StmtIdentifier> {
        match str.len() {
            // "type".len
            4 => {
                if str == b"type" {
                    Some(StmtIdentifier::SType)
                } else {
                    None
                }
            }
            // "interface".len == "namespace".len
            9 => {
                if str == b"interface" {
                    Some(StmtIdentifier::SInterface)
                } else if str == b"namespace" {
                    Some(StmtIdentifier::SNamespace)
                } else {
                    None
                }
            }
            // "abstract".len
            8 => {
                if str == b"abstract" {
                    Some(StmtIdentifier::SAbstract)
                } else {
                    None
                }
            }
            // "declare".len
            7 => {
                if str == b"declare" {
                    Some(StmtIdentifier::SDeclare)
                } else {
                    None
                }
            }
            // "module".len
            6 => {
                if str == b"module" {
                    Some(StmtIdentifier::SModule)
                } else {
                    None
                }
            }
            _ => None,
        }
    }

    pub static IMAP: phf::Map<&'static [u8], Kind> = phf::phf_map! {
        b"unique" => Kind::Unique,
        b"abstract" => Kind::Abstract,
        b"asserts" => Kind::Asserts,

        b"keyof" => Kind::PrefixKeyof,
        b"readonly" => Kind::PrefixReadonly,

        b"any" => Kind::PrimitiveAny,
        b"never" => Kind::PrimitiveNever,
        b"unknown" => Kind::PrimitiveUnknown,
        b"undefined" => Kind::PrimitiveUndefined,
        b"object" => Kind::PrimitiveObject,
        b"number" => Kind::PrimitiveNumber,
        b"string" => Kind::PrimitiveString,
        b"boolean" => Kind::PrimitiveBoolean,
        b"bigint" => Kind::PrimitiveBigint,
        b"symbol" => Kind::PrimitiveSymbol,

        b"infer" => Kind::Infer,
    };

    #[derive(Clone, Copy, PartialEq, Eq)]
    pub enum Kind {
        Normal,
        Unique,
        Abstract,
        Asserts,
        PrefixKeyof,
        PrefixReadonly,
        PrimitiveAny,
        PrimitiveNever,
        PrimitiveUnknown,
        PrimitiveUndefined,
        PrimitiveObject,
        PrimitiveNumber,
        PrimitiveString,
        PrimitiveBoolean,
        PrimitiveBigint,
        PrimitiveSymbol,
        Infer,
    }
}

#[derive(enumset::EnumSetType)]
pub enum SkipTypeOptions {
    IsReturnType,
    IsIndexSignature,
    AllowTupleLabels,
    DisallowConditionalTypes,
}

// PORT NOTE: Zig nested `Bitset` and `empty` inside `SkipTypeOptions`. Rust
// inherent associated types (`impl Foo { type Bar = ...; }`) are unstable
// (`inherent_associated_types`), so the alias and empty constant are hoisted
// to module scope.
pub type SkipTypeOptionsBitset = enumset::EnumSet<SkipTypeOptions>;
pub const SKIP_TYPE_OPTIONS_EMPTY: SkipTypeOptionsBitset = enumset::EnumSet::empty();

// ported from: src/js_parser/ast/TypeScript.zig
